src/Common/src/System/Net/Security/CertificateHelper.cs

```diff
@@ -33,39 +34,56 @@ internal static X509Certificate2 GetEligibleClientCertificate(X509Certificate2Co
                 return null;
             }
 
-            // Build a new collection with certs that have a private key. We need to do this manually because there is
-            // no X509FindType to match this criteria.
-            // Find(...) returns a collection of clones instead of a filtered collection, so do this before calling
-            // Find(...) to minimize the number of unnecessary allocations and finalizations.
-            var eligibleCerts = new X509Certificate2Collection();
             foreach (X509Certificate2 cert in candidateCerts)
             {
-                if (cert.HasPrivateKey)
+                if (!cert.HasPrivateKey)
                 {
-                    eligibleCerts.Add(cert);
+                    continue;
+                }
+                
+                if (IsValidClientCertificate(cert))
+                {
+                    return cert;
                 }
             }
 
-            // Don't call Find(...) if we don't need to.
-            if (eligibleCerts.Count == 0)
+            return null;
+        }
+
+        private static bool IsValidClientCertificate(X509Certificate2 cert)
+        {
+            foreach (X509Extension extension in cert.Extensions)
             {
-                return null;
+                if ((extension is X509EnhancedKeyUsageExtension eku) && !IsValidForClientAuthenticationEKU(eku))
+                {
+                    return false;
+                }
+                else if ((extension is X509KeyUsageExtension ku) && !IsValidForDigitalSignatureUsage(ku))
+                {
+                    return false;
+                }
             }
 
-            // Reduce the set of certificates to match the proper 'Client Authentication' criteria.
-            // Client EKU is probably more rare than the DigitalSignature KU. Filter by ClientAuthOid first to reduce
-            // the candidate space as quickly as possible.
-            eligibleCerts = eligibleCerts.Find(X509FindType.FindByApplicationPolicy, ClientAuthenticationOID, false);
-            eligibleCerts = eligibleCerts.Find(X509FindType.FindByKeyUsage, X509KeyUsageFlags.DigitalSignature, false);
+            return true;
+        }
 
-            if (eligibleCerts.Count > 0)
-            {
-                return eligibleCerts[0];
-            }
-            else
+        private static bool IsValidForClientAuthenticationEKU(X509EnhancedKeyUsageExtension eku)
+        {
+            foreach (Oid oid in eku.EnhancedKeyUsages)
```

Comment:
@bartonjs I used the original algorithm you gave me when i converted the .Find() API logic to explicit checks walking the X509Extensions.  How would I incorporate ApplicationPolicy extensions here?  Did the X509Certificate2Collection.Find() API do that already internally?

Reply:
There's CertificatePolicies, which is a different thing.  @morganbr, I'm not quite sure what you're referring to.  The Unix implementation of Find only does what this does.

Reply:
I'm referring to the property in the screenshot below since the name matches .NET's property. Unfortunately, both the underlying API and this property aren't well documented, so it's a bit hard to tell for sure what this is supposed to do.

![image](https://user-images.githubusercontent.com/11641665/40213411-2f1e5dd4-5a0a-11e8-93aa-0ea29853674b.png)

Reply:
This has some more information:
https://security.stackexchange.com/questions/30766/what-is-the-difference-between-extendedkeyusage-and-an-application-policy

>In Microsoft terminology, "application policies" appears to be a custom extension with OID 1.3.6.1.4.1.311.21.10, which seems to be roughly similar in contents and goals to the standard Extended Key Usage. I see in an MS-produced certificate an "Extended Key Usage" extension which contains two OID, and an "application policies" extension which contains the two same OID, albeit each within an extra SEQUENCE layer, which probably means that the Microsoft-specific extension allows for some optional qualifiers. I could not find a public description of the extension. (Edit: this page says that the extension uses the "same encoding as szOID_CERT_POLICIES", which is 2.5.29.32, i.e. the standard Certificate Policies extension from RFC 5280.)
Given the lack of information on this extension, the best way to handle it is probably to ignore it altogether. The extension is not marked "critical" so you are allowed to ignore it if you do not understand it, as per X.509, section 4.2:

As well as this:
https://www.sysadmins.lv/blog-en/certificate-policies-extension-all-you-should-know-part-1.aspx

Reply:
This gives more guidance:
https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2008-R2-and-2008/cc731792(v=ws.10)

>Application policies are sometimes called extended key usage or enhanced key usage. Because some implementations of public key infrastructure (PKI) applications cannot interpret application policies, both application policies and enhanced key usage sections appear in certificates issued by a Windows Server–based certification authority (CA).

This implies that checking the Application Policy extension would be redundant if we're already checking the Enhanced Key Usage extension looking for the absence (or presence) of the "Client Authentication" OID.

Reply:
My question would be whether this previously would find certificates that don't have the client auth oid in EKUs, but do in Application Policies and/or if it would exclude certificates that have the Application Policies extension, but don't have the client auth oid in it (because that may imply it's not intended for client auth).

Reply:
It sounds like only a Microsoft CA would even issue certificates with the Application Policies extension. And it is supposed to have the same information regarding the common OIDs (like Client Authentication) in both Application Policies and Enhanced Key Usage extensions. So, I don't think there would be variations in those between the two extensions.  The Application Policies extension also has other Microsoft specific OIDs in them as well.

Reply:
Based on this discussion, it doesn't seem like this PR is changing the behavior of how this method was working.  So, I plan to merge this PR.  However, if we get customer feedback about Application Policies extension that uses conflicting info compared with EKU extension, then we can file issues as needed on that.